lib/resolve.go

```diff
@@ -0,0 +1,138 @@
+package lib
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/qri-io/dataset"
+	"github.com/qri-io/qri/base"
+	"github.com/qri-io/qri/base/dsfs"
+	"github.com/qri-io/qri/dsref"
+	"github.com/qri-io/qri/fsi"
+)
+
+// ParseAndResolveRef combines reference parsing and resolution
+func (inst *Instance) ParseAndResolveRef(ctx context.Context, refStr, source string) (dsref.Ref, string, error) {
+	ref, err := dsref.Parse(refStr)
+
+	// bad case references are allowed-but-warned for backwards compatibility
+	if errors.Is(err, dsref.ErrBadCaseName) {
+		log.Error(dsref.ErrBadCaseShouldRename)
+		err = nil
+	} else if err != nil {
+		return ref, "", fmt.Errorf("%q is not a valid dataset reference: %w", refStr, err)
+	}
+
+	resolvedSource, err := inst.ResolveReference(ctx, &ref, source)
+	return ref, resolvedSource, err
+}
+
+// ResolveReference finds the identifier & HEAD path for a dataset reference.
+// the source parameter determines which subsystems of Qri to use when
+func (inst *Instance) ResolveReference(ctx context.Context, ref *dsref.Ref, source string) (string, error) {
+	if inst == nil {
+		return "", dsref.ErrNotFound
+	}
+
+	// Handle the "me" convenience shortcut
+	if ref.Username == "me" {
+		// TODO (b5) - this should be reading from a better place, and erroring if
+		// a canonical profile cannot be found for whatever reason
+		ref.Username = inst.cfg.Profile.Peername
+	}
+
+	resolvers, err := inst.resolveSources(source)
+	if err != nil {
+		return "", err
+	}
+
+	for _, resolver := range resolvers {
+		resolvedSource, err := resolver.ResolveRef(ctx, ref)
+		if err != nil {
+			if errors.Is(err, dsref.ErrNotFound) {
+				continue
+			} else {
+				return "", err
+			}
+		}
+
+		return resolvedSource, nil
+	}
+
+	return "", dsref.ErrNotFound
+}
+
+func (inst *Instance) resolveSources(source string) ([]dsref.Resolver, error) {
+	switch source {
```

Comment:
I think we're speaking to two different problems. To me the proposed code is great, and answers the problem of having fine grained control.

I wan to use this PR to answer some nagging issues of  inconsistent behaviour. Because users give us human reference strings, and we need to use both the network and local systems to get stable identifiers, we need a clear, enforceable behavior that's easy to test.

The problem is easier to see when at reference resolution over the network, where we want to measure both _internal_ consistency (all subsystems of a peer that retain references update with changes) and _network_ consistency (peers in the network agree on the same InitID). If we look at the networking parts of the codebase, lots of packages do some sort of reference resolution, they're just calling it by a different name:

* [`remote.ResolveHeadRef`](https://github.com/qri-io/qri/blob/2489d18cb091dbf381bf28acc259e9065316cbcc/remote/remote.go#L215-L224)
* [`remoteClient.ResolveHeadRefHTTP`](https://github.com/qri-io/qri/blob/2489d18cb091dbf381bf28acc259e9065316cbcc/remote/peer_sync_client.go#L247-L260)
* [`mockRemoteClient.ResolveHeadRef`](https://github.com/qri-io/qri/blob/2489d18cb091dbf381bf28acc259e9065316cbcc/remote/mock_client.go#L104-L107) 
* [`QriNode.RequestDataset`](https://github.com/qri-io/qri/blob/2489d18cb091dbf381bf28acc259e9065316cbcc/p2p/dataset.go#L16-L20)

There are more. If you broaden to include things like registry search (which implicitly resolves references), and logsync (which uses local resolution for comparison),
Many of these reference resolution functions are themselves compositions atop other subsystems, which if implemented incorrectly, will bubble up in strange ways. All of these are doing variations on the same behavior, inconsistently. We can't just _remove_ these because they're necessary. They're both ends of of a network call, or are a mock implementation intended to make testing easier.

By clearly defining an expected behaviour for any system that caches a reference, we can test that system for consistency against other subsystems of the same peer, and across the network. This is the real win that comes out of specified behavior: tests that make more sense.

I've updated this PR to include a spec test, an example of an internal consistency check (comparing dscache & logbook to confirm the hook is working), and a network consistency check (`lib.TwoActorRegistryIntegration` now checks all actors end in a consistent local state).